operator/pkg/ciliumendpointbatch/endpointbatch.go

```diff
@@ -0,0 +1,307 @@
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Cache CiliumEndpointStore Interface locally
+	c.ciliumEndpointStore = ces
+
+	// On operator warm boot, remove stale cep entries present in CEB
+	c.ciliumRemoveStaleCepEntries()
+
+	// Reconcile current state with desired state at every
+	// ciliumEndpointBatchUpsertSyncPeriod
+	log.WithFields(logrus.Fields{
+		"Upsert syncperiod": c.ciliumEndpointBatchUpsertSyncPeriod,
+		"Delete syncperiod": c.ciliumEndpointBatchDeleteSyncPeriod,
+	}).Debug("Starting CEB controller reconciler.")
+
+	go wait.Until(c.worker, c.ciliumEndpointBatchUpsertSyncPeriod, stopCh)
+
+	go func() {
+		defer utilruntime.HandleCrash()
+	}()
+
+	<-stopCh
+
+	return
+}
+
+// Upon warm boot[restart], Iterate over all CEPs which we got from the api-server
+// and compare it with CEPs packed inside CEB.
+// If there are any stale CEPs present in CEBs, remove them from their CEB.
+func (c *CiliumEndpointBatchController) ciliumRemoveStaleCepEntries() {
+	log.Info("Remove stale CEP entries in CEB")
+
+	// Get all ceps from local datastore
+	staleCeps := c.Manager.getAllCeps()
+
+	// Remove stale CEP entries present in CEB
+	for cepName, cCep := range staleCeps {
+		if _, exists, err := c.ciliumEndpointStore.GetByKey(getCepNameFromCCEP(cCep)); err == nil && exists || err != nil {
+			continue
+		}
+		log.WithFields(logrus.Fields{
+			"cep-name": cepName,
+		}).Debug("Removing stale CEP entry.")
+		c.Manager.RemoveCepFromCache(cCep)
+	}
+	return
+}
+
+// syncs all cebs from cebStore in manager cache.
+func syncCebsInLocalCache(cebStore cache.Store, manager cebManager) {
+
+	for _, obj := range cebStore.List() {
+		ceb := obj.(*v2alpha1.CiliumEndpointBatch)
+		// If CEB is already cached locally, do nothing.
+		if _, err := manager.getCebFromCache(ceb.GetName()); err == nil {
+			continue
+		}
+
+		// Create new CEB locally, with the given cebName
+		manager.createCeb(ceb.GetName())
+
+		// Deep copy the ceb, we got from api-server to local datastore.
+		manager.updateCebInCache(ceb, true)
+
+	}
+	log.Debug("Successfully synced all CEBs locally")
+	return
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same ceb
+// at the same time
+func (c *CiliumEndpointBatchController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *CiliumEndpointBatchController) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncCeb(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *CiliumEndpointBatchController) handleErr(err error, key interface{}) {
+
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.queue.Forget(key)
```

Comment:
Adding it back to `queue` wouldn't help, because we maxed out our retries and it will choke the system, eventually overwhelm the api-server. If there is any update to that CEB, it will be synced to api-server.

Reply:
What i meant is, in future if there is any update to that `CEB`,  it will be synced with api-server.

Reply:
I see, can you add a comment describing that we will ditch the update if it exceeds mas retries and that this is intentional?

Reply:
alright, i'll add comment.

Reply:
Done.

Reply:
Added TODO for adding metrics